google/blockly
# Movement Updates

## The basics




## The details
### Resolves


#2205 

### Proposed Changes



Adds a new "move" object configuration option with the following properties:
* scrollbars: Determines if the workspace has scrollbars. Defaults to true if the workspace has categories.
* drag: Determines if the workspace can be dragged with the mouse. Always false if scrollbars is false (at least in options parsing). Defaults to true if scrollbars is true.
* wheel: Determines if the workspace can be scrolled with the mouse wheel. Always false if scrollbars is false (at least in options parsing). Defaults to false.

Parsing does not currently allow these options to be independent, but you can make them independent by overriding the drag & wheel options after the workspace has been injected. This behavior is supported.

Changes the zoom function so that the zoom is "anchored" to the mouse position, even without scrollbars.

### Reason for Changes


People have requested scroll on wheel. #110

Also I think it may be useful to start separating the movement functionality from scrollbars to work towards #2145

### Test Coverage


  
Independent-ness of Scrollbars/Drag/Scrollwheel
1. Set move > scrollbars option to true. Set move > drag option to false. Set move > wheel option to false.
2. Observed that scrollbars alone worked. (pass)
3. Set move > scrollbars option to false. Overwrote move > drag to be true. Set move > wheel option to false. 
4. Observed that the dragging alone worked. And that scrollbars were not created. (pass)
5. Set move > scrollbars option to false. Set move > drag option to false. Overwrote move > wheel to be true.
6. Placed a block so I could observe the scrolling.
7. Observed that the scrollwheel alone worked. And that scrollbars were not created. (pass)

Scrolling and Zooming Interaction
1. Set move > scrollbars option to true. Set move > wheel option to false. Set zoom > wheel option to true.
2. Scrolled the mouse wheel. Observed how the workspace zoomed. (pass)
3. Set move > scrollbars option to true. Set move > wheel option to true. Set zoom > wheel option to true.
4. Scrolled the mouse wheel. Observed how the workspace scrolled vertical. (pass)
5. Hit shift and scrolled the mouse wheel. Observed how the workspace scrolled horizontally. (pass)
6. Hit ctrl and scrolled the mouse wheel. Observed how the workspace zoomed. (pass)
7. Set move > scrollbars option to true. Set move > wheel option to true. Set zoom > wheel option to false.
8. Scrolled the mouse wheel. Observed how the workspace scrolled. (pass)
9. Hit shift and scrolled the mouse wheel. Observed how the workspace scrolled horizontally. (pass)

Zooming With and Without Scrollbars
1. Set scrollbars option to true. Set zoom > wheel option to true. Placed a block to visualize zooming.
2. Scrolled the mouse wheel. Observed how the zoom was anchored to the mouse. (pass)
3. Pressed the zoom in control. Observed how it zoomed towards the center. (pass)
4. Changed scrollbars option to false.
5. Scrolled the mouse wheel. Observed how the zoom was anchored to the mouse. (pass)
6. Pressed the zoom in control. Observed how it zoomed towards the center. (pass)
7. Changed the zoom > wheel option to false. 
8. Pressed the zoom in control. Observed how it zoomed towards the top left corner. (pass)

Note: Steps 7&8 of the above make it so the content is no longer bounded. If you were to, for example, override drag to make the content bounded it would zoom towards the center.

Tested on:
* Desktop Chrome









### Additional Information


All of this works, but there are still some things I want to do (like clean up calls to scrollbars.set) so I consider it a little bit WIP. I just wanted to post a PR before I went too crazy doing things, so I can still change stuff easily.
